File: stats.py
from __future__ import annotations

from datetime import datetime, timezone
from time import monotonic
from typing import Any

from constants import DATA_DIR
from utils import json_load, json_save


STATS_PATH = DATA_DIR / "stats.json"
DEFAULTS: dict[str, Any] = {
    "drops_claimed": 0,
    "mining_minutes": 0,
    "channel_switches": 0,
    "watch_heartbeats": 0,
    "watch_failures": 0,
    "started_count": 0,
    "last_claim_at": "",
    "last_progress_at": "",
    "last_heartbeat_at": "",
}


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class Stats:
    """Tiny persistent counters plus per-process operational state."""

    def __init__(self) -> None:
        self.lifetime = json_load(STATS_PATH, DEFAULTS)
        self.session = {key: 0 for key in DEFAULTS if isinstance(DEFAULTS[key], int)}
        self.started_at = _now()
        self.started_monotonic = monotonic()
        self.last_inventory_at: str | None = None
        self.last_recovery_at: str | None = None
        self.lifetime["started_count"] += 1
        self.session["started_count"] = 1
        self.save()

    def increment(self, name: str, amount: int = 1, *, stamp: str | None = None) -> None:
        self.lifetime[name] = int(self.lifetime.get(name, 0)) + amount
        self.session[name] = int(self.session.get(name, 0)) + amount
        if stamp:
            self.lifetime[stamp] = _now()
        self.save()

    def progress(self, minutes: int) -> None:
        if minutes > 0:
            self.increment("mining_minutes", minutes, stamp="last_progress_at")

    def heartbeat(self, succeeded: bool) -> None:
        self.increment("watch_heartbeats", stamp="last_heartbeat_at")
        if not succeeded:
            self.increment("watch_failures")

    def claim(self) -> None:
        self.increment("drops_claimed", stamp="last_claim_at")

    def save(self) -> None:
        json_save(STATS_PATH, self.lifetime, sort=True)

    def snapshot(self) -> dict[str, Any]:
        return {
            "startedAt": self.started_at,
            "uptimeSeconds": int(monotonic() - self.started_monotonic),
            "session": dict(self.session),
            "lifetime": dict(self.lifetime),
            "lastInventoryAt": self.last_inventory_at,
            "lastRecoveryAt": self.last_recovery_at,
        }

File: src/stats.ts
/**
 * Port of `stats.py`: tiny persistent counters plus per-process state.
 *
 * `STATS_DEFAULTS` and the JSON layout are identical to Python, so a
 * `stats.json` written by either runtime loads in the other. Pass an
 * explicit `dataPath` (tests use a temp dir); production uses
 * `<TDM_DATA_DIR>/stats.json`.
 */

import { jsonLoad, jsonSave } from "./jsonStore.ts";

export const STATS_DEFAULTS: Record<string, number | string> = {
  drops_claimed: 0,
  mining_minutes: 0,
  channel_switches: 0,
  watch_heartbeats: 0,
  watch_failures: 0,
  started_count: 0,
  last_claim_at: "",
  last_progress_at: "",
  last_heartbeat_at: "",
};

export interface StatsSnapshot {
  startedAt: string;
  uptimeSeconds: number;
  session: Record<string, number>;
  lifetime: Record<string, number | string>;
  lastInventoryAt: string | null;
  lastRecoveryAt: string | null;
}

export class Stats {
  readonly lifetime: Record<string, number | string>;
  readonly session: Record<string, number>;
  readonly startedAt: string;
  private readonly startedMonotonic: number;
  lastInventoryAt: string | null = null;
  lastRecoveryAt: string | null = null;

  constructor(private readonly dataPath: string) {
    this.lifetime = jsonLoad(dataPath, { ...STATS_DEFAULTS });
    this.session = {};
    for (const [key, value] of Object.entries(STATS_DEFAULTS)) {
      if (typeof value === "number") this.session[key] = 0;
    }
    this.startedAt = new Date().toISOString();
    this.startedMonotonic = performance.now();
    this.lifetime["started_count"] = Number(this.lifetime["started_count"] ?? 0) + 1;
    this.session["started_count"] = 1;
    this.save();
  }

  increment(name: string, amount = 1, stamp?: string): void {
    this.lifetime[name] = Number(this.lifetime[name] ?? 0) + amount;
    this.session[name] = Number(this.session[name] ?? 0) + amount;
    if (stamp) this.lifetime[stamp] = new Date().toISOString();
    this.save();
  }

  progress(minutes: number): void {
    if (minutes > 0) this.increment("mining_minutes", minutes, "last_progress_at");
  }

  heartbeat(succeeded: boolean): void {
    this.increment("watch_heartbeats", 1, "last_heartbeat_at");
    if (!succeeded) this.increment("watch_failures");
  }

  claim(): void {
    this.increment("drops_claimed", 1, "last_claim_at");
  }

  save(): void {
    jsonSave(this.dataPath, { ...this.lifetime }, true);
  }

  snapshot(): StatsSnapshot {
    return {
      startedAt: this.startedAt,
      uptimeSeconds: Math.floor((performance.now() - this.startedMonotonic) / 1000),
      session: { ...this.session },
      lifetime: { ...this.lifetime },
      lastInventoryAt: this.lastInventoryAt,
      lastRecoveryAt: this.lastRecoveryAt,
    };
  }
}
